### ISSUE TITLE
Kiali assumes Grafana is cluster internal, in instio-system namespace and has k8s service named grafana

### ISSUE THREAD

**Issue by iftachsc** [kiali/kiali#865]
**Describe the bug**
I've set Grafana url, user , and password for an external grafana where we send all metrics.
kiali it self has all its metrics from a federated prometheus inside the cluster but we dont hold a grafana there.
everything works. kiali generates correct links and send me to grafana. only thing is that kiali still complain on welcome screen with the following:

![image](https://user-images.githubusercontent.com/28829551/53188883-ede28c00-360e-11e9-9ef0-ccb4fcd9b803.png)

**Versions used**
Kiali: 0.15
Istio: 1.0.3-gke.3
GKE

**To Reproduce**
i think its related ot any installation where you set grafana url, user and passowrd correctly but its external and doesnt sit in side istio-system and has k8s service named grafana

**Expected behavior**
if kiali ables to connect succesfully to the grafana set in grafana.url configuration - dont try to find a k8s service named grafana.

I currently fix this with defining a fake externalName k8s service with the ip of the grafana but i think i could set any ip there.

**Comment by jotak** [comment]
@iftachsc thanks for reporting, I'll have a look

**Comment by jotak** [comment]
@iftachsc , I have some questions for you. Actually I don't really understand how Kiali would generate correct links to grafana if it doesn't have access to its http API.

For instance the link displayed here:

![sshot](https://screenshotscdn.firefoxusercontent.com/images/95dbcce0-8108-4aaa-aa32-3f9cd2f30175.png)
Looks like this:
`http://(grafana_server)/d/UbsSZTDik/istio-workload-dashboard?var-namespace=myns&var-workload=myworkload`

The *UbsSZTDik* part is not hard-written, Kiali has to query grafana's rest api to get that dashboard ID.
Note that older versions of Grafana did not have this generated part.

So, I don't see how you can get valid links without that connection between Kiali and Grafana, and that connection is currently only performed with in-cluster queries. Do you have an explanation for that.

Doing the same with external Grafana would necessitate more plumbing with egress definition and so on, which isn't something that we cover now.

Maybe what we could do, however, is when a Grafana url is configured but is inaccessible from within kiali, have a fallback scenario to show that url but without pointing to the exact service/workload dashboard, as we cannot guess its ID.

**Comment by iftachsc** [comment]
i do have a connection. but kiali complains regardless. i have no issue between kiali and grafana. 
regardless i am not sure kiali really query grafana api - it just knows the templating of the istio dashbaords but any way thats not an issue.

**Comment by jotak** [comment]
@iftachsc can you show what the grafana url looks like from kiali? Something like the one I posted above?

**Comment by fcervantes2** [comment]
@jotak It shouldn't matter if Grafana is internal or external. Kiali should still be able to use Grafana's API regardless. As you mentioned, it does require plumbing between Kiali and Grafana, but that should not be the responsibility of Kiali. With that in mind, providing credentials/API keys, and URL (if different) to Grafana's REST API to Kiali should be sufficient.

**Comment by jotak** [comment]
@fcervantes2 I know it is technically possible, but I don't understand how Kiali would currently generate correct URL given how it is coded right now (url is currently generated after some queries that are performed only in cluster)... that's why I'm asking the question above.

**Comment by sfudeus** [comment]
I have a similar issue with kiali 0.19 and an external grafana installation for which I did configure a url in the config. I need to check on the comments above if I did something wrongly, but generally I have the same issue.

**Comment by sfudeus** [comment]
@xeviknal this is one of the items we discussed.

**Comment by jotak** [comment]
@sfudeus I've a couple of questions :
- Do you see some errors in Kiali logs about reaching the Grafana server?
- In metrics dashboards, do you see links to Grafana or not? Said differently, is the problem that these links are incorrect, or not there at all?
- Can you open the browser's javascript console and check in network tab what's the response to the query on `/api/grafana` ? (you may have to refresh the page to see this request)

**Comment by sfudeus** [comment]
@jotak thanks for the fast reply, I'm currently boarding the plane after leaving KubeCon, but I'll check your questions tomorrow 😊

**Comment by jotak** [comment]
@sfudeus sure, have a safe travel :-)

**Comment by jmazzitelli** [comment]
Is this related? https://github.com/istio/istio/issues/14335

**Comment by jotak** [comment]
@jmazzitelli yes maybe, that was my impression too.

**Comment by jotak** [comment]
@sfudeus also one thing to know is if you have this issue in a cluster environment with grafana and the kube cluster not on the same host, or if you're using say minikube + grafana on same machine.

**Comment by Demonian** [comment]
@jotak In my situation I have grafana and kiali in the same cluster, but grafana is deployed to custom namespace (not `istio-system').
cc @jmazzitelli

**Comment by sfudeus** [comment]
@jotak 
The only seemingly relevant errors in the logfile I see are the following: 

```
* E0524 18:54:32.463232       1 grafana.go:43] services "grafana" not found
* E0524 18:56:22.027245       1 dashboards.go:375] Runtimes discovery failed, cannot load 
dashboards in namespace ams-istio-dev. Error was: unknown (get monitoringdashboards.monitoring.kiali.io)
```

The first one correlates to the error message in the UI:
```
Error fetching Grafana Info., Error: [ services "grafana" not found ]
24.5.201920:54:32
```
when accessing the "Inbound Metrics" of a Service, no special errors are logged, just some prometheus queries.

There are no links at all to Grafana. 

Calls to /api/grafana return error 503 (```{"error":"services \"grafana\" not found"}```)

The following is my config.yaml for kiiali:
```
    istio_namespace: istio-system
    server:
      port: 20001
    external_services:
      istio:
        url_service_version: http://istio-pilot:8080/version
      jaeger:
        url: "https://jaeger-ui.k8s-fe-sandbox-iz1-bs.poinfra.server.lan/"
        service: jaeger-query.jaeger-system
      grafana:
        url: https://<my-grafana-host>
```

Defining an empty service entry for grafana just leads to a different error message (```{"error":"resource name may not be empty"}```)

And regarding my cluster setup: 
Istio and kiali are running in multi-node bare-metal cluster and grafana is located outside of kubernetes as a standalone installation. 

My versions: 
Kiali 0.9, Istio 1.1.6, Prometheus: 2.7.2, K8s: 1.14.1

**Comment by jotak** [comment]
@sfudeus I sent a pull request that (I hope) should address your issue: #1096 . So you would have now to specify `in_cluster: false` in Kiali config for Grafana.

@Demonian have you tried to just set the namespace for grafana in Kiali config? I think if you just write something like this, it should work:
```yaml
# ...
  external_services:
    grafana:
        service: "grafana"	
        service_namespace: "the-grafana-namespace"
        url: https://my-grafana-url.com.ua
# ...
```

**Comment by xeviknal** [comment]
@sfudeus if that works for you, you will have it available next release: June 8th. Thanks for the feedback.

btw, I remember we talked about other bugs you had. it would be great if you have time to add issues for that. Something like the expansion node, one ux problem with the filters, filtering namespaces by labels on overview page and so on. 
Do not hesitate to reach us for anything. you would help the community so much. We will be happy to help :)

**Comment by sfudeus** [comment]
@jotak Looks better now, thx, I'm now getting an unknown authority error (no surprise). Where would I inject an additional trusted root ca into the docker image? 
Bonus question: Would I be able to add basic auth credentials to the URL?

@xeviknal Will do, thanks. I saw #212, which goes in the direction of the namespace filtering partially.

**Comment by jotak** [comment]
@sfudeus there's a ticket to allow untrusted / self-signed certificates, if it's what you want : https://github.com/kiali/kiali/issues/786#issuecomment-498147693
It's already on my top priority list.

About basic auth for Grafana, it should be possible to configure Kiali: see https://github.com/kiali/kiali/issues/689 . In the Kiali config you can set `username`, `password` and `api_key` (see https://github.com/kiali/kiali/blob/master/operator/deploy/kiali/kiali_cr.yaml#L141-L155 )
Since this is not something that we test every day, please let us know if anything is broken.

**Comment by jotak** [comment]
PR that should work with self-signed certificates (but I haven't fully tested yet) : #1112

**Comment by sfudeus** [comment]
@jotak The state of master works for me now (with in_cluster: false and insecure_skip_verify: true). No error message in ui anymore and a working link to grafana is shown from e.g. the workload metrics page. Thanks a lot. 
It would be ideal if I needn't resort to insecure_skip_verify but could just add my own CA. In my case the certificate is not self-signed but signed by an internal corporate CA.

**Comment by jotak** [comment]
@sfudeus there's ongoing work about that : https://github.com/kiali/kiali/pull/1129 (currently only for prometheus but the same will be done for grafana)
If you want to take a look and confirm it would also work for you ...

**Comment by jotak** [comment]
Hi @sfudeus 
The aforementioned PRs have been merged, so I imagine the config you need now should look more or less like that for off-cluster w/ basic auth:

```yaml
# ...
  external_services:
    grafana:
      in_cluster_url: "" # you can leave it empty
      url: "https://..."
      auth:
        type: "basic"
        ca_file: "..." # path to file in pod
        password: "..."
        username: "..."
```

**Comment by sfudeus** [comment]
Hi @jotak , yes thanks, we are running on 0.21 already and I adapted the config similarly as you suggested. Currently I still have set insecure_skip_verify in auth, ideally I manage to get it done with the ca_file. Ill report back as soon as I tested this. It seems the integration works fine already without specific credentials, since only the connectivity is checked. It seems no call is made which must be authenticated (at least in our setup).
